File: leds_unity/Assets/LevelsManager.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System;

public class LevelsManager : MonoBehaviour
{
    List<LevelData> levels;
    List<LevelZone> allZones;

    int levelID;
    int areaID;
    int numLeds;
    float frameRate;
    int lastLength = 150;
    string lastStatus = "";

    public Data allData;
    [Serializable]
    public class Data
    {
        public List<AreaData> areas;
    }
    [Serializable]
    public class AreaData
    {
        public List<ZoneData> zones;
        public List<LevelData> levels;
        public int length;
    }
    [Serializable]
    public class ZoneData
    {
        public string color;
        public float pos; // normalized (0, 1)
    }
    [Serializable]
    public class LevelData
    {
        public string status; //"safe": zona libre con transparencia // "status": "full" hace daño
        public float speed;
        public int initialLength;
        public int nextLength;
        public int seconds;
        public string ease;
    }
    public void Init(int numLeds, float frameRate, string json)
    {
        this.frameRate = frameRate;
        allZones = new List<LevelZone>();
        this.numLeds = numLeds;
        SetLevels(json);
        AddNewArea();
    }
    void SetLevels(string json)
    {
        allData = JsonUtility.FromJson<Data>(json);
        Debug.Log(json);
        levels = new List<LevelData>(); 
    }
    void AddLevel(int nextLength, float speed, int seconds, string status, string ease)
    {
        nextLength = nextLength * (numLeds / 100); //numLeds/100 normaliza de 0 a 100 el scalesss
        LevelData lData = new LevelData();
        lData.speed = speed;
        lData.initialLength = lastLength;
        lData.ease = ease;

        if (status == "")
            lData.status = lastStatus;
        else
            lData.status = status;

        lData.nextLength = nextLength;
        lData.seconds = seconds;
        levels.Add(lData);
        
        lastStatus = status;
        lastLength = nextLength;
    }
    
    void AddNewArea()
    {
        lastLength = 0;
        Debug.Log("AddNewArea: " + areaID);
        AreaData areaData = allData.areas[areaID];
        allZones = new List<LevelZone>();
        float qty = areaData.zones.Count;
        for (int a = 0; a < qty; a++)
        {
            int length = areaData.length * (numLeds / 100); //numLeds/100 normaliza de 0 a 100 el scalesss
            LevelZone levelzone = new LevelZone();
            Color color = GetColor(areaData.zones[a].color);
            levelzone.Init(a, this, areaData.zones[a].pos*numLeds, length, color, numLeds, frameRate);
            allZones.Add(levelzone);
        }
        levels = new List<LevelData>();
        foreach (LevelData lData in areaData.levels)
        {
            AddLevel(lData.nextLength, lData.speed, lData.seconds, lData.status, lData.ease);
        }
        activeLevelData = levels[0];
        areaID++;
    }
    Color GetColor(string colorName)
    {
        switch (colorName)
        {
            case "red": return Color.red;
            case "green": return Color.green;
            case "blue": return Color.blue;
        }
        return Color.black;
    }
    public void OnUpdate(float deltaTime)
    {
        foreach (LevelZone level in allZones)
            UpdateLevel(level, deltaTime);
    }
    void UpdateLevel(LevelZone level, float deltaTime)
    {
     //   Debug.Log("area: " + areaID + "   Level " + levelID + "   nextLength" +  activeLevelData.nextLength + "   initialLength:" +  activeLevelData.initialLength + "    speed:" + activeLevelData.speed + "    seconds:" + activeLevelData.seconds + " level " + level.from + " to:" + level.to);

        if (activeLevelData.nextLength != activeLevelData.initialLength)
            level.ScaleTo(activeLevelData.nextLength, activeLevelData.seconds, activeLevelData.ease, deltaTime);
        if (activeLevelData.speed != 0)
            level.Move(activeLevelData.speed, activeLevelData.seconds, activeLevelData.ease, deltaTime);

        level.Process(activeLevelData.seconds, activeLevelData.status, deltaTime);
    }
    LevelData activeLevelData;
    public void OnNextLevel()
    {
        Debug.Log("OnNextLevel  area: " + areaID + "   Add Level " + levelID);
        levelID++;
        if (levelID > levels.Count - 1)
        {
            levelID = 0;
            AddNewArea();
        }
        else
        {
            activeLevelData = levels[levelID];
        }
        foreach (LevelZone levelzone in allZones)
            levelzone.Restart(activeLevelData.initialLength);
    }
    public List<LevelZone> GetLevelZones()  { return allZones; }
    
}

File: micropython/base/LevelsManager.py
import levelZone
import json
        
d ='''{
  "areas":
  [
      {
        "length": "0",
        "zones":
        [
            {
              "color": "",
              "pos": ""
            }
        ],
        "levels":
        [
            {
              "nextLength": 0,
              "speed": 0,
              "seconds": 0,
              "ease": 0
            }
        ]
      }
  ]
}'''

levels = [object]
allZones = [object]

framerate = 0
levelID = 0
areaID = 0
numLeds = 0
lastLength = 150
lastStatus = ""

class LevelData:
    status = ""#; //"safe": zona libre con transparencia // "status": "full" hace daño
    speed = 0.0
    initialLength = 0
    nextLength = 0
    seconds = 0
    ease = ""

with open('data.json') as s:
    d = json.load(s)
    #print(d)
    

def Init(_numLeds, _frameRate):
    global numLeds
    frameRate = _frameRate
    allZones = []
    numLeds = _numLeds
    AddNewArea(frameRate, numLeds)

def AddLevel(_nextLength, speed, seconds, status, ease):
    global lastLength
    global lastStatus
    global numLeds
    global levels
    
    nextLength = _nextLength * (numLeds / 100) # numLeds/100 normaliza de 0 a 100 el scalesss
    print("level nextLength", _nextLength, " speed: ", speed, " seconds:", seconds, " status", status, " ease:", ease, " numLeds: ", numLeds)
    lData = LevelData
    lData.speed = speed
    lData.initialLength = lastLength
    lData.nextLength = nextLength
    lData.seconds = seconds
    lData.status = status
    lData.ease = ease

    if status == "":
        lData.status = lastStatus
    else:
        lData.status = status

    lData.nextLength = nextLength
    lData.seconds = seconds
    levels = []
    levels.append(lData)
    
    lastStatus = status
    lastLength = nextLength


def AddNewArea(framerate, numLeds):
    global allZones
    global areaID
    global thisLevelZone
    global levels
    
    areaData = d["areas"][areaID]
    length = areaData["length"] * (numLeds / 100)#//numLeds/100 normaliza de 0 a 100 el scalesss
    lastLength = length
    zones = areaData["zones"]
    
    
    allZones = [object]
    
    zN = len(zones)
    
    _id = 0
    for a in range(zN):               
        thisLevelZone = levelZone.LevelZone()
        zonesData = zones[a]
        
        if len(zonesData)>0:
            color = GetColor(zonesData["color"])
            pos = zonesData["pos"]*numLeds
        else:
            color = (0,0,0)
            pos = 0
           
        thisLevelZone.Init(_id, pos, length, color, numLeds, framerate)
        allZones.append(thisLevelZone)
        _id = _id+1
        
    levels = [object]
    l = areaData["levels"]
    lN = len(l)
    print("Add area id:", areaID, " zones: ", zN, " levels:", lN, " numLeds:", numLeds)
      
    for a in range(lN):        
        lData = l[a]
        speed = 0
        status = ""
        nextLength = 0
        seconds = ""
        ease = ""
        
        if lData.get("status"):
            status = lData["status"]
        if lData.get("speed"):
            speed = lData["speed"]
        if lData.get("nextLength"):
            nextLength = lData["nextLength"]
        if lData.get("seconds"):
            seconds = lData["seconds"]
        if lData.get("ease"):
            ease = lData["ease"]
            
        AddLevel(nextLength, speed, seconds, status, ease)    
    
    areaID = areaID+1
    if areaID > len(d["areas"])-1:
        areaID = 0

def GetColor(colorName):
    if colorName == "red":
        return (255,0,0)
    elif colorName == "green":
        return (0,255,0)
    elif colorName == "blue":
        return (0,0,255)
    else:
        return (0,0,0)

def OnUpdate(deltaTime):
    total = len(allZones)
    for a in range(total):
        if a < len(allZones):
            level = allZones[a]
            UpdateLevel(level, deltaTime)

def UpdateLevel(level, deltaTime):
    activeLevelData = levels[levelID]
    #print("initialLength: ", activeLevelData.initialLength,  " nextLength", activeLevelData.nextLength,  " seconds", activeLevelData.seconds  )
       
       
    if activeLevelData.nextLength != activeLevelData.initialLength:
        level.ScaleTo(activeLevelData.nextLength, activeLevelData.seconds, activeLevelData.ease, deltaTime)
    if activeLevelData.speed != 0:
        level.Move(activeLevelData.speed, activeLevelData.seconds, activeLevelData.ease, deltaTime)

    level.Process(activeLevelData.seconds, activeLevelData.status, deltaTime)

def OnNextLevel():
    global levelID
    global activeLevelData
    global framerate
    global numLeds
    levelID = levelID+1
    
    if levelID > len(levels) - 1:    
        levelID = 0
        AddNewArea(framerate, numLeds)
    else:    
        activeLevelData = levels[levelID]
        
    activeLevelData = levels[levelID]
    for a in range(len(allZones)):
        LevelZone = allZones[a]
        LevelZone.Restart(activeLevelData.initialLength)

def GetLevelZones():
    return allZones 
